In [1]:
# Import Packages
import sys, os
import pandas as pd
import geopandas as gpd
import seaborn as sns
import numpy as np
import matplotlib.pyplot as plt
import math
import json
import plotly.graph_objects as go
import plotly.express as px
from plotly.offline import iplot
from scipy import stats
from ipyleaflet import (Map, GeoData, basemaps, WidgetControl, GeoJSON,
 LayersControl, Icon, Marker,basemap_to_tiles, Choropleth,
 MarkerCluster, Heatmap,SearchControl, 
 FullScreenControl, LayerGroup, LegendControl)
from ipywidgets import Text, HTML, widgets, interactive, HBox, VBox
from branca.colormap import linear
%matplotlib inline

# Size and Extent of Fairfax County's Municipal Separate Storm Sewer (MS4)

Fairfax County's MS4 permit requires the County to map the `MS4 service area` and each `MS4 outfall`. Wood has identified over 10,000 terminal outfalls in the County, of which over 7,000 are the County's responsibility. The drainage areas to each MS4 outfall are delineated using an automated approach. This approach uses a hydrologically conditioned Digital Elevation Model (DEM) that incorporates StormNet to enforce subsurface drainage networks. This process is revisited annually to incorporate updates to StormNet and is documented in further detail in a SOP.

**MS4 outfall**
: the terminal point of manmade infrastructure that discharges to a water of the state or water of the US. Manmade infrastructure includes earthen ditches maintained to convey stormwater flow. 

**<font color=green>All MS4 drainage areas</font>**
: All the area draining to each MS4 outfall.

**<font color=red>MS4 service area</font>**
: Area draining to each County MS4 outfall, _except_ for excluded lands such as:
- VPDES permittees
- Other MS4 permittees (ex. VDOT)
- Forested lands (Note: the County chose to _not exclude_ Forested Lands from the MS4 service area)

The map below shows three layers/geometries that the County can be evaluated on:
1. <font color=red>MS4 service area</font>
2. <font color=green>All MS4 drainage areas</font>
3. <font color=blue>The entire County</font>

In [2]:
Drainage_Areas = gpd.read_file("Data/All_Drainage_Areas.geojson", driver="GeoJSON")
Drainage_Areas = Drainage_Areas.rename(columns={"STORMNET_I":"STORMNET_ID"})

In [3]:
Service_Areas = gpd.read_file("Data/County_MS4_Service_Areas_Sim1.geojson", driver="GeoJSON")
Service_Areas = Service_Areas.rename(columns={"STORMNET_I":"STORMNET_ID"})

In [4]:
County = gpd.read_file("https://opendata.arcgis.com/datasets/58cf8abd870e47aeb1be8911983d2d44_15.geojson")


*Note: Gif below displays the difference between 4 relevant geometries - <font color=green>All Drainage Areas</font>, County Drainage Areas, the <font color=red>MS4 Service Area</font>, and the <font color=blue>Entire County</font>.*

In [5]:
from IPython.display import Image, display
Image(url='https://media.giphy.com/media/XSICRHtziK2jKnlrwr/giphy.gif')

In [6]:
center = (38.8554638, -77.2757340)
m = Map(center=center, zoom=10)

geodata_SA = GeoData(geo_dataframe = Service_Areas,
                  style={'color': 'black', 'fillColor': '#eb1b0c', 'opacity':0.03, 'weight':1.9, 'dashArray':'2', 'fillOpacity':0.6},
                         hover_style={'fillColor': '#b08a3e', 'fillOpacity': 0.8},
                                      name = 'MS4 Service Area')
geodata_DA = GeoData(geo_dataframe = Drainage_Areas,
                  style={'color': 'black', 'fillColor': '#2deb0c', 'opacity':0.03, 'weight':1.9, 'dashArray':'2', 'fillOpacity':0.6},
                         hover_style={'fillColor': '#b08a3e', 'fillOpacity': 0.8},
                                      name = 'All Drainage Areas')
geodata_County = GeoData(geo_dataframe = County,
                  style={'color': 'black', 'fillColor': '#053ef7', 'opacity':0.03, 'weight':1.9, 'dashArray':'2', 'fillOpacity':0.2},
                         hover_style={'fillColor': '#b08a3e', 'fillOpacity': 0.8},
                                      name = 'County')

m.add_layer(geodata_SA)
m.add_layer(geodata_DA)
m.add_layer(geodata_County)

search = SearchControl(position="topleft", 
                       url=
                       'https://nominatim.openstreetmap.org/search?format=json&q={s}',
                       zoom=14,
                       property_name='display_name'
                      )
m.add_control(search)
control = FullScreenControl()
m.add_control(control)
m.add_control(LayersControl())
legend = LegendControl({"MS4 Service Area":"#eb1b0c", "All Drainage Areas":"#2deb0c", "County":"#053ef7"}, name="Legend", position="bottomright")
m.add_control(legend)

m

Map(center=[38.8554638, -77.275734], controls=(ZoomControl(options=['position', 'zoom_in_text', 'zoom_in_title…

In [7]:
MS4_Scenarios = pd.read_csv("Data/MS4_Scenarios.csv")
MS4_Scenarios = MS4_Scenarios.iloc[0:3]

The chart below depicts three stacked bars. Each bar corresponds to a polygon geometry displayed in the map above. Within each geometry (ex. MS4 service area ~ current MS4) the area can be broken down into three categories:
1. Impervious Area
2. Pervious Area
3. Forest within the Pervious Area

*Note: Excluded Lands (other than forest) have been removed from each geometry*

In [8]:
MS4_Scenarios_AC = MS4_Scenarios[['Scenarios','Impervious Area', 'Non-Forest Pervious Area', 'Forest in Pervious Area']]
MS4_Scenarios_AC = pd.melt(MS4_Scenarios_AC, id_vars = ['Scenarios'])

In [9]:
fig_bar = px.bar(MS4_Scenarios_AC, x='Scenarios', y='value', color='variable', color_discrete_sequence =['red','blue','green'],
             labels={
                     "Scenarios": "MS4 Responsibility Scenarios",
                     "value": "Acres",
                     "variable": "Land Cover Type"
                 },
             title='MS4 With Excluded Lands Removed')

f1 = go.FigureWidget(fig_bar)
f1

FigureWidget({
    'data': [{'alignmentgroup': 'True',
              'hovertemplate': ('Land Cover Type=Imperv…

In [10]:
MS4_Scenarios[['Scenarios','Impervious Area', 'Non-Forest Pervious Area', 'Forest in Pervious Area']]

Scenarios Impervious Area  \
0                                       Current MS4       21,012.45   
1  Drainage To All County Outfalls Of Any Ownership       42,867.46   
2                                     Entire County       56,494.71   

  Non-Forest Pervious Area Forest in Pervious Area  
0                20,951.35               29,887.81  
1                33,377.83               44,194.92  
2               134,446.57               69,381.93

The next chart displays the `Excluded Lands`. Chesapeake Bay TMDL Special Condition Guidance Memo GM20-2003 specifices that permittees may exclude from their service area, in addition to lands regulated by a separate MS4 permit, the following areas:

1. Land regulated under any General VPDES permit that addresses industrial stormwater, including the General VPDES Permit for Stormwater Associated with Industrial Activity (VAR05), the General VPDES Permit for Concrete Products Facilities (VAG11), and the Nonmetallic Mineral Processing General Permit (VAR84);
2. Lands regulated under an individual VPDES permit for industrial stormwater discharges;
3. Forested Lands (the County chose **not** to remove these lands);
4. Agricultural Lands;
5. Wetlands; and,
6. Open Waters.

Each bar on this chart also corresponds with a geometry; however, the geometry is slightly different than before. Instead of using the MS4 service area for the "Current MS4", we use County drainage areas. The County drainage areas do not have excluded lands removed - such as the VDOT MS4/ROW. 

Each bar shows the excluded lands present within the specified geometry.


In [11]:
MS4_Scenarios_Removed = pd.read_csv("Data/MS4_Scenarios_Removed.csv")

In [12]:
MS4_Scenarios_Removed_melt = pd.melt(MS4_Scenarios_Removed, id_vars = ['Scenarios'], value_vars = ['VDOT', 'Other MS4s', 'VPDES Permitees'])

In [13]:
fig_bar1 = px.bar(MS4_Scenarios_Removed_melt, x='Scenarios', y='value', color='variable', color_discrete_sequence =['orange','blue','red'],
             labels={
                     "Scenarios": "MS4 Responsibility Scenarios",
                     "value": "Acres",
                     "variable": "Potentially Excluded Lands"
                 },
             title='Excluded Lands')

f2 = go.FigureWidget(fig_bar1)
f2

FigureWidget({
    'data': [{'alignmentgroup': 'True',
              'hovertemplate': ('Potentially Excluded L…

In [14]:
MS4_Scenarios_Removed

Scenarios         VDOT   Other MS4s  \
0                                       Current MS4  12973.30102   369.190364   
1  Drainage To All County Outfalls Of Any Ownership  18793.31833  1878.809257   
2                                     Entire County  24483.20377  4747.520370   

   VPDES Permitees  
0       311.178562  
1       970.206672  
2      3462.219218

The County's MS4 permit will require it to reduce 100% of the L2 Scoping Run POC reductions required for existing sources as of June 30, 2009 by the end of its third permit cycle. Table 3b from the Chesapeake Bay TMDL Guidance can be used to estimate the scope of these required reductions within the Potomac River Basin. 

Impervious and pervious area estimates for the three geometries can be plugged into the column`Land served by MS4 (2009)` to estimate what reductions would be required under different MS4 size and extent scenarios. The results are shown in the interactive chart below. 

The dropdown menu allows the user to select different POC.

### Table 3b. Current MS4
#### Calculation Sheet for Estimating Existing Source Loads and Reduction Requirements for the Potomac River Basin

| Pollutant | Subsource | Loading rate (lbs/ac/yr) | Land served by MS4 (2009) | Load (lbs/yr) | L2 Required Reduction | 100% Cumulative Reduction Required (lbs/yr)
| :---------: | ----------- | :---------: | :---------: |  :---------: |  :---------: | :---------: |
| TN | Regulated impervious | 16.86 | 21,012.45 | 354,269.91 | 9% | 31,884.29
| TN | Regulated pervious |10.07 | 50,839.16 | 511,950.34 | 6% | 30,717.02
| TP | Regulated impervious | 1.62 | 21,012.45 | 34,040.17 | 16% | 5,446.43
| TP | Regulated pervious |0.41 | 50,839.16 | 20,844.06 | 7.25% | 1,511.19
| TSS | Regulated impervious | 1,171.32 | 21,012.45 | 24,612,302.93 | 20% | 4,922,460.59
| TSS | Regulated pervious |175.8 | 50,839.16 | 8,937,524.33| 7.25% | 782,033.38

In [15]:
MS4_Scenarios_POC = MS4_Scenarios[['Scenarios','TN', 'TP', 'TSS']]

In [16]:
MS4_Scenarios_POC = MS4_Scenarios_POC.set_index(['Scenarios'])

In [17]:
fig = go.Figure()

for column in MS4_Scenarios_POC.columns.to_list():
    fig.add_trace(
        go.Bar(
            x = MS4_Scenarios_POC.index,
            y = MS4_Scenarios_POC[column],
            name = column
        )
    )
  
fig.update_layout(
    updatemenus=[go.layout.Updatemenu(
        active=0,
        buttons=list(
            [dict(label = 'All',
                  method = 'update',
                  args = [{'visible': [True, True, True]},
                          {'title': 'All',
                           'showlegend':True}]),
            dict(label = 'TN',
                  method = 'update',
                  args = [{'visible': [True, False, False]},
                          {'title': 'TN',
                           'showlegend':True}]),
             dict(label = 'TP',
                  method = 'update',
                  args = [{'visible': [False, True, False]}, # the index of True aligns with the indices of plot traces
                          {'title': 'TP',
                           'showlegend':True}]),
             dict(label = 'TSS',
                  method = 'update',
                  args = [{'visible': [False, False, True]},
                          {'title': 'TSS',
                           'showlegend':True}]),
            ])
        )
    ])

f2 = go.FigureWidget(fig)
f2

#fig.show()

FigureWidget({
    'data': [{'name': 'TN',
              'type': 'bar',
              'uid': 'b351617a-c915-49…

In [18]:
MS4_Scenarios_POC

TN         TP  \
Scenarios                                                                 
Current MS4                                        62,601.31   6,957.62   
Drainage To All County Outfalls Of Any Ownership  111,916.53  13,417.09   
Entire County                                     208,878.25  20,702.23   

                                                            TSS  
Scenarios                                                        
Current MS4                                        5,704,493.97  
Drainage To All County Outfalls Of Any Ownership  11,235,564.70  
Entire County                                     16,370,068.97

The table below displays the County's current Chesapeake Bay TMDL `credit ledger` for all eligible practices for each *Pollutant of Concern (POC)*. This `credit ledger` is a summation of work to date and can be compared to the required reductions for each of the three geometries. 

In [19]:
Progress_Data = pd.read_csv("Data/Progress_Data.csv")
Progress_Data.set_index('Practice')

TN        TP         TSS
Practice                                                 
Structural Retrofits        6906.81    664.51   772500.27
Stream Restoration         35665.31  11015.48  3420425.95
2006-2009                   5705.48    670.27   577628.02
Redevelopment                177.30     24.61    13751.44
SFR 2009-2014                135.32     22.09    11977.54
SFR Post 2014                843.81     81.86    15254.22
In-Lake Forebay Retrofits   6615.89    391.88   156038.24
Septic Conversion           2148.33      0.00        0.00
NMPs                          89.58      1.82        0.00
Land Use Change              202.16     12.15     5645.84
Other                          0.00      0.00        0.00
Off-Site Credits           17276.00   1361.00        0.00

The graph below compares the County's `credit ledger` with the required reductions for the:

1. <font color=red>Current MS4 Service Area</font>
2. <font color=green>All Drainage Areas</font>
3. <font color=blue>Entire County</font>

It calculates the `Percent Towards Goal` for each eligible practice and summarizes the total for each *POC*. The goal is assumed to be 100% compliance, even though the County is currently only required to meet 40% compliance for the total required reductions. The three geometries are broken out into three facets, which show progress towards compliance under each scenario.

In [20]:
Progress_Data = Progress_Data.assign(Current_MS4_TN_Perc= lambda Progress_Data: (Progress_Data.TN/62601.31)*100).round(2)
Progress_Data = Progress_Data.assign(Current_MS4_TP_Perc= lambda Progress_Data: (Progress_Data.TP/6957.62)*100).round(2)
Progress_Data = Progress_Data.assign(Current_MS4_TSS_Perc= lambda Progress_Data: (Progress_Data.TSS/5704493.97)*100).round(2)

In [21]:
Progress_Data_1 = Progress_Data[['Practice','Current_MS4_TN_Perc', 'Current_MS4_TP_Perc', 'Current_MS4_TSS_Perc']]
Progress_Data_1 = Progress_Data_1.rename(columns={"Current_MS4_TN_Perc": "TN", "Current_MS4_TP_Perc":"TP", "Current_MS4_TSS_Perc":"TSS"})
Progress_Data_1 = pd.melt(Progress_Data_1, id_vars=['Practice'])
Progress_Data_1['Scenario'] ='Current MS4'

In [22]:
Progress_Data_2 = Progress_Data
Progress_Data_2 = Progress_Data_2.assign(TN_Perc= lambda Progress_Data: (Progress_Data.TN/111916.53)*100).round(2)
Progress_Data_2 = Progress_Data_2.assign(TP_Perc= lambda Progress_Data: (Progress_Data.TP/13417.09)*100).round(2)
Progress_Data_2 = Progress_Data_2.assign(TSS_Perc= lambda Progress_Data: (Progress_Data.TSS/11235564.70)*100).round(2)

In [23]:
Progress_Data_2 = Progress_Data_2[['Practice','TN_Perc', 'TP_Perc', 'TSS_Perc']]
Progress_Data_2 = Progress_Data_2.rename(columns={"TN_Perc": "TN", "TP_Perc":"TP", "TSS_Perc":"TSS"})
Progress_Data_2 = pd.melt(Progress_Data_2, id_vars=['Practice'])
Progress_Data_2['Scenario'] ='All Drainage Areas'

In [24]:
Progress_Data_3 = Progress_Data
Progress_Data_3 = Progress_Data_3.assign(TN_Perc= lambda Progress_Data: (Progress_Data.TN/208878.25)*100).round(2)
Progress_Data_3 = Progress_Data_3.assign(TP_Perc= lambda Progress_Data: (Progress_Data.TP/20702.23)*100).round(2)
Progress_Data_3 = Progress_Data_3.assign(TSS_Perc= lambda Progress_Data: (Progress_Data.TSS/16370068.97)*100).round(2)

In [25]:
Progress_Data_3 = Progress_Data_3[['Practice','TN_Perc', 'TP_Perc', 'TSS_Perc']]
Progress_Data_3 = Progress_Data_3.rename(columns={"TN_Perc": "TN", "TP_Perc":"TP", "TSS_Perc":"TSS"})
Progress_Data_3 = pd.melt(Progress_Data_3, id_vars=['Practice'])
Progress_Data_3['Scenario'] ='Entire County'

In [26]:
Progress_Plot_Data = pd.concat([Progress_Data_1, Progress_Data_2, Progress_Data_3])
Progress_Plot_Data = Progress_Plot_Data.rename(columns={"variable": "Pollutants", "value":"Percent Towards Goal"})

In [31]:
fig_h = px.bar(Progress_Plot_Data, x="Percent Towards Goal", y="Pollutants", color='Practice', facet_row = "Scenario", color_discrete_sequence=["blue","orange","silver","yellow","green","black","pink","brown","white","light green","light blue","red"], orientation='h',
             hover_data=["Practice", "Percent Towards Goal"],
             height=800,
             title='Current Progress').update_yaxes(categoryorder="category descending")
fig_h.add_vline(x=100, line_width=2)
fig_h.add_vline(x=40, line_width=2, line_dash="dash")
f_h = go.FigureWidget(fig_h)
f_h

FigureWidget({
    'data': [{'alignmentgroup': 'True',
              'customdata': array([['Structural Retrofi…